src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/QRScanner/ExposureSubmissionQRScannerViewModel.swift

```diff
@@ -83,27 +153,58 @@ class ExposureSubmissionQRScannerViewModel: NSObject, AVCaptureMetadataOutputObj
 	/// - the guid is a well formatted string (6-8-4-4-4-12) with length 43
 	///   (6 chars encode a random number, 32 chars for the uuid, 5 chars are separators)
 	func extractGuid(from input: String) -> String? {
-		guard
-			!input.isEmpty,
-			input.count <= 150,
-			let regex = try? NSRegularExpression(
-				pattern: "^https:\\/\\/localhost\\/\\?(?<GUID>[0-9A-Fa-f]{6}-[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12})$"
-			),
-			let match = regex.firstMatch(in: input, options: [], range: NSRange(location: 0, length: input.utf8.count))
-		else { return nil }
+		guard !input.isEmpty,
+			  input.count <= 150,
+			  let urlComponents = URLComponents(string: input),
+			  urlComponents.scheme == "https",
+			  urlComponents.host == "localhost",
+			  let candidate = urlComponents.query,
+			  candidate.count == 43,
+			  let matchings = candidate.range(
+				of: #"^[0-9A-Fa-f]{6}-[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12}$"#,
+				options: .regularExpression
+			  ) else {
+			return nil
+		}
+		return matchings.isEmpty ? nil : candidate
+	}
 
-		guard let range = Range(match.range(withName: "GUID"), in: input) else { return nil }
+	// MARK: - Private
 
-		let candidate = String(input[range])
-		guard !candidate.isEmpty, candidate.count == 43 else { return nil }
+	private let onSuccess: (DeviceRegistrationKey) -> Void
+	private let onError: (QRScannerError, _ reactivateScanning: @escaping () -> Void) -> Void
+	private let captureDevice: AVCaptureDevice?
 
-		return candidate
-	}
+	private(set) var isScanningActivated: Bool
 
-	// MARK: - Private
+	private func setupCaptureSession() {
+		guard let currentCaptureDevice = captureDevice,
+			let caputureDeviceInput = try? AVCaptureDeviceInput(device: currentCaptureDevice) else {
+			onError(.cameraPermissionDenied) { Log.error("Failed to setup AVCaptureDeviceInput", log: .ui) }
+			return
+		}
 
-	private var isScanningActivated: Bool
+		let metadataOutput = AVCaptureMetadataOutput()
+		captureSession.addInput(caputureDeviceInput)
+		captureSession.addOutput(metadataOutput)
+		metadataOutput.metadataObjectTypes = [.qr]
+		metadataOutput.setMetadataObjectsDelegate(self, queue: .main)
+	}
 
-	private let onSuccess: (DeviceRegistrationKey) -> Void
+	private func activateScanning() {
+		if isScanningActivated {
```

Comment:
Maybe you could use captureSession.isRunning instead of an own state ?

Reply:
this is only need for testing purpose